components/javascript-wrapper.js

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * This file contacts the common code used by all of the variations of any RDF
+ * node componentin the noflo implementation of the RDF Pipeline.
+ */
+
+var _ = require('underscore');
+var stateComponent = require('./state-component');
+
+/**
+ * This module provides a Javascript wrapper for updaters.  It is responsible for setting up the
+ * code to ensure updater callback, and then invoking the RDF pipeline framework to complete building
+ * the component and setting up the vni and state metadata.
+ *
+ * @param def updater definition. This can be a simple updater function, or an object that contains 
+ *        updater metadata (e.g., input/output port definitions, description, icons), and an updater.
+ *
+ * @return a promise to create the noflo rdf component
+ */
+module.exports = function(def) { 
+
+    var self = this;
+
+    if ( _.isUndefined(def) || ( _.isEmpty(def) && ! _.isFunction( def ))) { 
+        // Nothing to work with - no metadata or updater - can't proceed.
+        throw new Error('javascript-wrapper found no updater definition');
+    }
+
+    if ( _.isUndefined( def.updater ) && !_.isFunction(def)) { 
+       // got called with def metadata, but no updater for this one
+        return stateComponent( def );
+    }
+     
+    // Have an updater - figure out the updater arguments 
+    var updaterArgs = ( _.isFunction(def) ) ? introspect( def ) : introspect( def.updater );
+
+    // Do we have any input port definition?  If we were only passed an updater function, we might not.
+    // In that case, build the list of input port names from the updater function arguments.  We will 
+    // assume that all arguments are input ports.
+    if ( _.isFunction( def ) ) { 
+        def = { updater: def, inPorts: updaterArgs };
+    }
+
+    // set up an onchange callback so we can invoke the updater when we have changes to process
+    var updaterCallback = function() { 
+        return function( state ) {
+
+            // Do we have any updaterArgs?  We might not if we do not have source code
+            if ( _.isEmpty( updaterArgs ) ) { 
+               // No updaterArgs, so just pass the state
+               def.updater.call( self, state );
+
+            } else {
+                // Have one or more parameters - get them in the right order and pass them 
+                var parameters = _.map( updaterArgs, function( arg ) { 
+                    var parameter = state[arg]; 
+                    if ( _.isUndefined(parameter) && def.outPorts && def.outPorts.hasOwnProperty( arg ) ) {
+                        // passing in the name of the output port - is this correct? 
+                        parameter = arg;
+                    }
+                    return parameter; 
+                });
+
+                def.updater.apply( self, parameters );
```

Comment:
I'm surprised to see 'def' still used when the updater is called, because I think of def as only being used during graph creation -- not at runtime when the pipeline is sending events and processing data.  I think of it as only a compiletime structure.

Also, the updater needs a way to indicate whether or not it actually modified the node's state, so that the framework can know whether to send events downstream if it did.  The API for this was shown on slides 5-9 of http://dbooth.org/2013/pipeline/RDF-Pipeline-Implementation.pdf , but if we switch to a one-state-per-updater model then we may want to change this API.  But somehow the updater needs to be able to indicate that it did or did not end up changing the state.